cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,292 @@
+            a) `cirq.Gate` subclass
+            b) `cirq.Gate` instance
+            c) `cirq.GateFamily` instance
+
+        `cirq.Gate` subclasses and instances are converted to the default
+        `cirq.GateFamily(gate=g)` instance and thus a default name and
+        description is populated.
+
+        Args:
+            gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
+            `cirq.GateFamily` instances to initialize the Gateset.
+            name: (Optional) Name for the Gateset. Useful for description.
+        """
+        self._name = name
+        self._gates = frozenset(
+            [g if isinstance(g, GateFamily) else GateFamily(gate=g) for g in gates]
+        )
+
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
+    @property
+    def gates(self) -> FrozenSet[GateFamily]:
+        return self._gates
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        """Check for containment of `item` in any of the member `GateFamily`.
+
+        This method just forwards containment checks to the underlying
+        `GateFamily` objects. It returns `True` if any of the member gate
+        families can accept `item`.
+
+        To validate whether a `cirq.CircuitOperation`/`cirq.OP_TREE`/
+        `cirq.Circuit` is made up entirely of gates from this Gateset, use the
+        `validate` method given below.
+
+        Args:
+            item: The `cirq.Gate` or `cirq.Operation` instance to check
+                containment for.
+        """
+        for gate_family in self.gates:
+            if item in gate_family:
+                return True
+        return False
+
+    def validate(
+        self,
+        circuit_or_optree: Union['cirq.AbstractCircuit', op_tree.OP_TREE],
+        *,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
+    ) -> bool:
+        """Validates gates forming `circuit_or_optree` should be contained in Gateset.
+
+        Args:
+            circuit_or_optree: The `cirq.Circuit` or `cirq.OP_TREE` to validate.
+            unroll_circuit_op: If True, `cirq.CircuitOperation` is recursively
+                validated by validating the underlying `cirq.Circuit`.
+            accept_global_phase: If True, `cirq.GlobalPhaseOperation` is accepted.
+        """
+        # To avoid circular import.
+        from cirq.circuits import circuit
+
+        optree = circuit_or_optree
+        if isinstance(circuit_or_optree, circuit.AbstractCircuit):
+            optree = circuit_or_optree.all_operations()
+        return all(
+            self._validate_operation(
+                op, unroll_circuit_op=unroll_circuit_op, accept_global_phase=accept_global_phase
+            )
+            for op in op_tree.flatten_to_ops(optree)
+        )
+
+    def _validate_operation(
+        self,
+        op: raw_types.Operation,
+        *,
+        unroll_circuit_op: bool = True,
```

Comment:
When would you want this to be false ?

Reply:
Google backend supports serializing circuit operations and accepting circuits with CircuitOperations, but not all vendors would necessarily support them. Hence, during compilation, turning off the flag would mean that CircuitOperations are not accepted as already present in the target gateset and would continue to decompose into more primitive operations whose serialization is supported.

Reply:
Makes sense.